internal/gatewayapi/translator.go

```diff
@@ -0,0 +1,515 @@
+package gatewayapi
+
+import (
+	"fmt"
+
+	"github.com/envoyproxy/gateway/pkg/ir"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	KindGateway   = "Gateway"
+	KindHTTPRoute = "HTTPRoute"
+	KindService   = "Service"
+)
+
+// Cache is an interface that the Translator requires
+// for getting Gateway API and related resources.
+type Cache interface {
+	ListGateways() []*v1beta1.Gateway
+	ListHTTPRoutes() []*v1beta1.HTTPRoute
+	GetNamespace(name string) *v1.Namespace
+	GetService(namespace, name string) *v1.Service
+}
+
+// Translator translates Gateway API resources to the IR,
+// and computes status for Gateway API resources.
+type Translator struct {
+	gatewayClassName v1beta1.ObjectName
+}
+
+type TranslateResult struct {
+	Gateways   []*v1beta1.Gateway
+	HTTPRoutes []*v1beta1.HTTPRoute
+	IR         *ir.Xds
+}
+
+func newTranslateResult(gateways []*GatewayContext, httpRoutes []*HTTPRouteContext, xdsIR *ir.Xds) *TranslateResult {
+	translateResult := &TranslateResult{
+		IR: xdsIR,
+	}
+
+	for _, gateway := range gateways {
+		translateResult.Gateways = append(translateResult.Gateways, gateway.Gateway)
+	}
+	for _, httpRoute := range httpRoutes {
+		translateResult.HTTPRoutes = append(translateResult.HTTPRoutes, httpRoute.HTTPRoute)
+	}
+
+	return translateResult
+}
+
+func (t *Translator) Translate(cache Cache) *TranslateResult {
+	xdsIR := &ir.Xds{}
```

Comment:
expects `Name` to be set

Reply:
Any suggestions on what this should be set to? I didn't see where it was used.

Reply:
youre right, I'll rm it in a different PR. It might not really be needed here since Envoy/IR is always tied to a single GW Class, and we plan on passing all that info via a separate `Metadata` field in the future